dali/imgcodec/deferred_result.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <deque>
+#include <mutex>
+#include <condition_variable>
+#include "dali/imgcodec/image_decoder_interfaces.h"
+
+namespace dali {
+namespace imgcodec {
+
+class DeferredDecodeResults::Impl {
+ public:
+  static Impl *get() {
+    if (free_.empty())
+      return new Impl();
+
+    Impl *ret = free_.back().release();
+    free_.pop_back();
+    return ret;
+  }
+
+  static void put(Impl *impl) {
+    free_.emplace_back(impl);
+  }
+
+  void wait() {
+    if (results_ready_ == results_.size())
+      return;
+    std::unique_lock lock(mtx_);
+    cv_.wait(lock, [&]() {
+      return results_ready_ == results_.size();
+    });
+  }
+
+  void set(int index, DecodeResult res) {
```

Comment:
I'd also add a `set_all`. Imagine a decoder that always get the results for the whole batch at once.